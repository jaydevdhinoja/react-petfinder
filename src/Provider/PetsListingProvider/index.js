import React, { createContext, useContext, useState, useEffect } from 'react';
import { Client } from "@petfinder/petfinder-js";

const defaultState = {
	pets: [],
	searchTerm: ''
};

export const PetsListingContext = createContext(defaultState);

const client = new Client({apiKey: 'lZrY2wFLzhyEUefSPPn4JjEwIUBMP0gppdb1EEI9CXRRT629HL', secret: 'PgC6ma6wEX2Ll79BbODrGTX6IcKvwMnj7zVuAlOu'});

const PetsListingProvider = ({ children }) => {
	const InitialState = () => defaultState;

	const [ initialState, setInitialState ] = useState(InitialState());

	useEffect(
		() => {
			const loadPetsData = async () => {
				try {
					const animalType = {type: 'Dog' }
					const breed = initialState.searchTerm !== '' && initialState.searchTerm ? { breed: initialState.searchTerm } : {} 
					client.animal.search({...animalType, ...breed})
					.then(function (response) {
						let petfinderData = response.data;
						console.log(petfinderData)
						setInitialState({
							...initialState,
							pets: petfinderData.animals
						});
					})
					.catch(function (error) {
						console.log(error)
						setInitialState({
							...initialState, pets: []
						})
					});
				} catch (e) {
					console.error(e);
				}
			};
			loadPetsData();
		},
		[ initialState.searchTerm ]
	);

	return (
		<PetsListingContext.Provider
			value={{
				...initialState,
				setSearchTerm: (item) => {
					setInitialState({ ...initialState, searchTerm: item });
				}
			}}
		>
			{children}
		</PetsListingContext.Provider>
	);
};

export default PetsListingProvider;

export const usePetsListingContext = () => {
	return useContext(PetsListingContext);
};
