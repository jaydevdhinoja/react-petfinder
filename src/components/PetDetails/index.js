import React from "react";
import { usePetsListingContext } from '../../Provider/PetsListingProvider';
import { withRouter } from 'react-router-dom';

const PetDetails = (props) => {

  const { pets } = usePetsListingContext();

  const selectedPet = pets.length > 0 ? pets.find(pet => parseInt(pet.id) === parseInt(props.match.params.id)) : undefined;

  const imgSrc = () => {
    if(selectedPet.photos.length > 0) {
      return selectedPet.photos[0].large;
    } else {
      return process.env.PUBLIC_URL + '/DogPlaceholder.svg'
    }
}

  if(!selectedPet) {
    return (
      <div className="app_box app_container">
        Loading...
      </div>
    )
  } else {
    return (
      <div>
        <div className="app_back" onClick={() => props.history.push('/')}>Back to listing</div>
        <div className="app_box app_container">
          <div className="app_card_view">
            <div className="app_card_view_image">
                <img src={imgSrc()} alt={selectedPet.name}/>
            </div>
            <div className="app_card_view-details">
                <h2>{selectedPet.name}</h2>
                <p className='app_card-details'>{selectedPet.contact.email}</p>
                <p className='app_card-details'>{selectedPet.description}</p>
            </div>
          </div>
        </div>
      </div>
    );
  }
};

export default withRouter(PetDetails);
