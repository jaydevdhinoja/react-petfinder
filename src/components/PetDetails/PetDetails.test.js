import '@testing-library/jest-dom/extend-expect';
import React from 'react';
import PetDetails from '.';
import { PetsListingContext } from '../../Provider/PetsListingProvider';
import { render, cleanup } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom'
import { Router } from 'react-router-dom'
import { createMemoryHistory } from 'history'

afterEach(cleanup);

const InitialState = {
	pets: [
		{
			id: 47782140,
			organization_id: "FL850",
			url: "https://www.petfinder.com/dog/fred-47782140/fl/miami-beach/dachshund-rescue-south-florida-fl850/?referrer_id=b5f7363e-4d07-4756-ab82-d3a8060feccd",
			type: "Dog",
			species: "Dog",
			breeds: {primary: "Dachshund", secondary: null, mixed: false, unknown: false},
			colors: {primary: "Red / Chestnut / Orange", secondary: null, tertiary: null},
			age: "Adult",
			gender: "Male",
			size: "Small",
			name: "Fred",
			coat: "Short",
			description: "Can you say &quot;Personality? This is Fred and he has is in loads!  He is a handsome 9 year old...",
			photos: [
				{
					small: "https://dl5zpyw5k3jeb.cloudfront.net/photos/pets/47782140/5/?bust=1586432382&width=100",
					medium: "https://dl5zpyw5k3jeb.cloudfront.net/photos/pets/47782140/5/?bust=1586432382&width=300",
					large: "https://dl5zpyw5k3jeb.cloudfront.net/photos/pets/47782140/5/?bust=1586432382&width=600",
					full: "https://dl5zpyw5k3jeb.cloudfront.net/photos/pets/47782140/5/?bust=1586432382"
				}
			]
		}
	],
	searchTerm: ''
};

const defaultProps = {
	match: { params: { id: 47782140 }, path:'/:id' }
};

function renderWithContext(node, { value, ...options }) {
	return render(<PetsListingContext.Provider value={value}>{node}</PetsListingContext.Provider>, options);
}

function renderWithRouter(
	ui,
	{
	  route = '/',
	  history = createMemoryHistory({ initialEntries: [route] }),
	} = {}
  ) {
	const Wrapper = ({ children }) => (
	  <Router history={history}>{children}</Router>
	)
	return {
		...renderWithContext(render(ui, { wrapper: Wrapper })),
	  // adding `history` to the returned utilities to allow us
	  // to reference it in our tests (just try to avoid using
	  // this to test implementation details).
	  history,
	}
  }

describe('Pet Details', () => {
	test('renders component', () => {
		const { container } = renderWithContext(<PetDetails />, { value: InitialState, wrapper: MemoryRouter });
		expect(container).toMatchSnapshot();
	});

	test('renders component no pets', () => {
		const { getByText } = renderWithContext(<PetDetails />, {
			value: { pets:[], searchTerm:'' },
			wrapper: MemoryRouter
		});

		expect(getByText(/Loading.../i)).toBeInTheDocument();
	});
});
