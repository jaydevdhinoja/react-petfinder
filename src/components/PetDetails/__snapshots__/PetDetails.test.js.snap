// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pet Details renders component 1`] = `
<div>
  <div
    class="app_box app_container"
  >
    Loading...
  </div>
</div>
`;
